packages/test-sync/src/androidAndroidTest/kotlin/io/realm/kotlin/test/mongodb/shared/ProgressListenerTests.kt

```diff
@@ -0,0 +1,249 @@
+package io.realm.kotlin.test.mongodb.shared
+
+import io.realm.kotlin.Realm
+import io.realm.kotlin.ext.query
+import io.realm.kotlin.internal.platform.runBlocking
+import io.realm.kotlin.mongodb.User
+import io.realm.kotlin.mongodb.sync.Direction
+import io.realm.kotlin.mongodb.sync.Progress
+import io.realm.kotlin.mongodb.sync.ProgressMode
+import io.realm.kotlin.mongodb.sync.SyncConfiguration
+import io.realm.kotlin.mongodb.syncSession
+import io.realm.kotlin.test.mongodb.TEST_APP_PARTITION
+import io.realm.kotlin.test.mongodb.TestApp
+import io.realm.kotlin.test.mongodb.createUserAndLogIn
+import io.realm.kotlin.types.ObjectId
+import io.realm.kotlin.types.RealmObject
+import io.realm.kotlin.types.annotations.PrimaryKey
+import kotlinx.coroutines.async
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.drop
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.last
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.scan
+import kotlinx.coroutines.flow.takeWhile
+import kotlinx.coroutines.supervisorScope
+import kotlinx.coroutines.withTimeout
+import kotlin.test.AfterTest
+import kotlin.test.BeforeTest
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertFailsWith
+import kotlin.test.assertTrue
+import kotlin.test.fail
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.seconds
+
+private const val TEST_SIZE = 5000
+private val TIMEOUT = 10.seconds
+
+private val schema = setOf(ProgressListenerTests.ProgressTestObject::class)
+
+class ProgressListenerTests {
+
+    private lateinit var app: TestApp
+    private lateinit var partitionValue: String
+
+    @BeforeTest
+    fun setup() {
+        app = TestApp(appName = TEST_APP_PARTITION)
+        partitionValue = ObjectId.create().toString()
+    }
+
+    @AfterTest
+    fun tearDown() {
+        if (this::app.isInitialized) {
+            app.close()
+        }
+    }
+
+    @Test
+    fun downloadProgressListener_changesOnly() = runBlocking {
+        val uploadRealm = Realm.open(createSyncConfig(app.createUserAndLogIn()))
+
+        // Verify that we:
+        // - get a "transferComplete" event
+        // - complete the flow, and
+        // - that all objects are available afterwards
+        Realm.open(createSyncConfig(app.createUserAndLogIn())).apply {
```

Comment:
The `use { ... }` exposes the realm as `it` whereas `apply { ... }.close()` makes it the implicit receiver (`this`). I found a bit annoying when writing the tests, since most of the inner operations where done on the realm, but I can revert if people like that better.

Reply:
The advantage of using `use { }` over `apply { }.close()` is that the former correctly closes the Realm in case the test fails, so I would probably prefer that. (the outer Realm would also benefit from this)

You could use `with(realm.syncSession) { }` inside the test instead?